### PR TITLE
[BEAM-9770] Add BigQueryIO deadletter pattern

### PR DESCRIPTION
**Pull request by rezarokni**
[BEAM-9770] Add BigQueryIO deadletter pattern

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: examples/java/src/main/java/org/apache/beam/examples/snippets/Snippets.java [hunk 1]
```diff
@@ -753,35 +750,93 @@ public static void main(String[] args) {
...
+                                                new TableSchema()
+                                                        .setFields(
+                                                                com.google.common.collect.ImmutableList.of(
+                                                                        new TableFieldSchema()
+                                                                                .setName("num")
+                                                                                .setType("INTEGER")
+                                                                                .setMode("REQUIRED"))))
+                                        .to("Test.dummyTable")
+                                        .withFormatFunction(x -> new TableRow().set("num", (x == 2) ? null : x))
+                                        .withFailedInsertRetryPolicy(InsertRetryPolicy.neverRetry())
```
Comment (chamikaramj) [inline examples/java/src/main/java/org/apache/beam/examples/snippets/Snippets.java:809]: Probably change to "retryTransientErrors" ? I'm afraid many users will just copy and use this pipeline as a template and will keep this entry in their production pipelines.

File: website/src/documentation/patterns/bigqueryio.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+-->
+
+# Google BigQuery patterns
+
+The samples on this page show you common patterns for use with BigQueryIO.
+
+## BigQueryIO deadletter pattern
+In production systems, it is useful to implement the deadletter pattern with BigQueryIO outputting any elements which had errors during processing by BigQueryIO into another PCollection for further processing. 
+
+When using `STREAMING_INSERTS`  you can use the `WriteResult` object to access a `PCollection` with the `TableRows` that failed to be inserted into BigQuery.
```
Comment (chamikaramj) [inline website/src/documentation/patterns/bigqueryio.md:28]: Let's mention retry policies here and mention that that dictates which failures will be added to the deadletter queue and whether we will retry.

File: examples/java/src/main/java/org/apache/beam/examples/snippets/Snippets.java (hunk 1 above)
Comment (rezarokni) [reply examples/java/src/main/java/org/apache/beam/examples/snippets/Snippets.java]: Fixed.

File: website/src/documentation/patterns/bigqueryio.md (hunk 2 above)
Comment (rezarokni) [reply website/src/documentation/patterns/bigqueryio.md:28]: Fixed.

**Comment by chamikaramj** [comment]
LGTM. Thanks.

**Comment by rezarokni** [comment]
Run Java PreCommit

**Comment by rezarokni** [comment]
Run Java PreCommit

**Comment by rezarokni** [comment]
Run Website PreCommit

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by aaltay** [comment]
Python parts LGTM as well. I can merge this after tests pass.

**Comment by rezarokni** [comment]
Retest this please

**Comment by aaltay** [comment]
Retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
Run Java PreCommit

**Comment by aaltay** [comment]
Run Portable_Python PreCommit

**Comment by aaltay** [comment]
Tests, including lint are failing.

**Comment by rezarokni** [comment]
Looks like fails in CasandraIO which is odd as those files are not touched by this. will fix the lint issues...

**Comment by rezarokni** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by rezarokni** [comment]
Local tests ran ok, but also raised:
https://issues.apache.org/jira/browse/BEAM-10003

So this PR is now just the code bits , rather than code + website bits.

**Comment by rezarokni** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by rezarokni** [comment]
Merge got rid of changes to snippits.java adding that back

**Comment by rezarokni** [comment]
added back the stuff removed by merge  , so hopefully this can now go in. And then I will raise PR num two for the website changes.

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by aaltay** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by rezarokni** [comment]
org.apache.beam.runners.spark.SparkPortableExecutionTest.testExecution
org.apache.beam.fn.harness.FnHarnessTest.testLaunchFnHarnessAndTeardownCleanly

Flaky tests?

**Comment by aaltay** [comment]
Run Java PreCommit

**Comment by aaltay** [comment]
Run Java PreCommit

**Comment by aaltay** [comment]
Run Java PreCommit

**Comment by aaltay** [comment]
Run Java PreCommit

**Pull request merged**